google-cloud-core/src/main/java/com/google/cloud/Policy.java

```diff
@@ -158,64 +174,134 @@ protected Builder(Policy policy) {
       checkNotNull(bindings, "The provided map of bindings cannot be null.");
+      checkArgument(
+          !isConditional(this.version, this.bindingsList),
+          "setBindings() is only supported with version 1 policies and non-conditional policies");
       for (Map.Entry<Role, Set<Identity>> binding : bindings.entrySet()) {
         checkNotNull(binding.getKey(), "The role cannot be null.");
         Set<Identity> identities = binding.getValue();
         checkNotNull(identities, "A role cannot be assigned to a null set of identities.");
         checkArgument(!identities.contains(null), "Null identities are not permitted.");
       }
-      this.bindings.clear();
+      // convert into List format
+      this.bindingsList.clear();
       for (Map.Entry<Role, Set<Identity>> binding : bindings.entrySet()) {
-        this.bindings.put(binding.getKey(), new HashSet<>(binding.getValue()));
+        Binding.Builder bindingBuilder = Binding.newBuilder();
+        bindingBuilder.setRole(binding.getKey().getValue());
+        for (Identity identity : binding.getValue()) {
+          bindingBuilder.addMembers(identity.strValue());
+        }
+        this.bindingsList.add(bindingBuilder.build());
+      }
+      return this;
+    }
+
+    /**
+     * Replaces the builder's List of bindings with the given List of Bindings.
+     *
+     * @throws NullPointerException if the given map is null or contains any null keys or values
+     * @throws IllegalArgumentException if any identities in the given map are null
+     */
+    public final Builder setBindings(List<Binding> bindings) {
+      for (Binding binding : bindings) {
+        checkNotNull(binding.getRole(), "The role cannot be null.");
+        List<String> members = binding.getMembers();
+        checkNotNull(members, "A role cannot be assigned to a null set of identities.");
+        checkArgument(!members.contains(null), "Null identities are not permitted.");
+      }
+      this.bindingsList.clear();
+      for (Binding binding : bindings) {
+        Binding.Builder bindingBuilder = Binding.newBuilder();
+        bindingBuilder.setMembers(ImmutableList.copyOf(binding.getMembers()));
+        bindingBuilder.setRole(binding.getRole());
+        bindingBuilder.setCondition(binding.getCondition());
+        this.bindingsList.add(bindingBuilder.build());
       }
       return this;
     }
 
     /** Removes the role (and all identities associated with that role) from the policy. */
     public final Builder removeRole(Role role) {
-      bindings.remove(role);
+      checkArgument(
+          !isConditional(this.version, this.bindingsList),
+          "removeRole() is only supported with version 1 policies and non-conditional policies");
+      for (int i = 0; i < bindingsList.size(); ++i) {
+        Binding binding = bindingsList.get(i);
+        if (binding.getRole().equals(role.getValue())) {
+          bindingsList.remove(i);
```

Comment:
we shouldn't remove from the list within a for loop. Have you considered using an iterator?

Reply:
I had a messy time with this specific case, do you have an example using an iterator?